### PR TITLE
Add tests for local/non-local diagnostic classification for accessor block

### LINKED ISSUE #67310: Diagnostic reported on property symbol name shouldn't be classified as a non-local diagnostic
Issue identified here: https://github.com/dotnet/roslyn/pull/67298/files#diff-be8c569a25b5f64e947b8469fd4f3ec5ff50c0b04d09b59dff822e7f1f7a4edfR477-R494

1. @sharwell's PR #67298 has identified an actual bug. The IDE test that had to be skip non-local diagnostic validation never causes the lightbulb to show up in the editor.
2. The core issue is highlighted here: https://github.com/dotnet/roslyn/blob/65cd3dcc15df7fb39a138efc0faea84acd467816/src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs#L96-L98. For symbols with block body, `SyntaxReference.GetSyntax()` gets the syntax node for the entire symbol decl + the body. While for symbols with expression body, `SyntaxReference.GetSyntax()` seems to return only the `ArrowExpressionClauseSyntax`. The analyzer is itself working around this case in the code above.
3. We need to adjust the driver code to account for the case where operation blocks/code blocks for expression bodies consider the parenting PropertyDeclarationSyntax/IndexerDeclarationSyntax as topmost node for analysis, so that the diagnostic reported for this case is considered a local diagnostic. This issue tracks updating analyzer driver to handle this case and consider such a diagnostic as a local diagnostic.

### PR DESCRIPTION
**Pull request by mavasani**
Closes #67310

Add tests to assert current local/non-local diagnostic classification for accessor blocks

~~When analyzing code blocks/operation blocks corresponding to expression bodies, consider the parenting PropertyDeclarationSyntax/IndexerDeclarationSyntax/MethodDeclarationSyntax as the topmost node for analysis and diagnostic reporting. This allows analyzers to report local diagnostics on the these declaration syntax nodes when analyzing code blocks/operation blocks.~~

~~Verified that both the product changes in this PR are required for the added unit test to pass~~

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5188]: 1. what happens for other accessors (like setters)?
2. PropertyDeclarationSyntax/IndexerDeclarationSyntax share a common BasePropertyDeclarationSyntax, can you use that?
3. why only for property/indexer.  Is there a concern about add/remove for events?
4. why only for property/indexer?  What about methods/constructors/destructors/operarators?

If hte answer is: it works for all those other cases, can we add tests?

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5186]:
```suggestion
                    if (declaringSyntax is ArrowExpressionClauseSyntax { Parent: BasePropertyDeclarationSyntax })
```

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5188]: oh, i see why this isn't for events. this is not for the accessor case, but when you have `int P => ...`  still unsure about all the other method/constructor/destructor/operator cases.

File: src/Compilers/CSharp/Test/Emit2/Diagnostics/GetDiagnosticsTests.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Emit2/Diagnostics/GetDiagnosticsTests.cs:745]: tests for constructors, destructors, operators, accessors (property, indexer, event plz).

**Comment by mavasani** [comment]
> Can you also update `MakeStructMemberReadOnlyTests` to remove the use of `CodeFixTestBehaviors.SkipLocalDiagnosticCheck`?

Yes, I'll do so once your PR is merged.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5188]: I also don't understand this.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5188]: so the only time when you'd have a `=>` parented by a BaseProperty is if you have `int Foo => ...` or `int this[...] => ...`.  In either case, the ArrayExpr corresponds to the GetAccessor and nothing else (that's how i understand the code).

Comment (mavasani) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5188]: > so the only time when you'd have a => parented by a BaseProperty is if you have int Foo => ... or int this[...] => .... In either case, the ArrayExpr corresponds to the GetAccessor and nothing else (that's how i understand the code).

That is correct.

However, I have put some more thought into this, and I think we want to allow analyzers to report diagnostics anywhere within the Property/Indexer declaration sub-tree from operation/node callbacks inside accessory body, regardless of it being an expression body or a block body. For example, if an analyzer that gets a callback for literal `0` inside accessor body and reports a diagnostic on the identifier `P`, it would be weird if the diagnostic is classified as a local diagnostic for expression body case, but not for block body case. I believe we want both the cases to be treated uniformly.

I am going to update the logic, add more tests for expression body and block body for all method kinds: accessors, ctors, dtors, operators, ordinary methods, etc. that validate all of them have identical behavior.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5188]: @mavasani that makes a lot of sense to me :)

Comment (mavasani) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5188]: I also figured out that the current implementation in the PR will change the Node/Operation passed into CodeBlockStart/OperationBlockStart callbacks. That would be a breaking change, which we definitely want to avoid. I am going to change the implementation a bit to avoid this and instead just adjust the scope for local/non-local diagnostic classification. I'll also add this validation in the added unit tests.

Comment (mavasani) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: I have pushed an updated implementation.

**Comment by mavasani** [comment]
> Oh, is there also a concern for local functions?

Local functions are just regular statements within the code block/operation block. So, they need no special handling.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5191]: Nit:
```suggestion
            if (symbol is IMethodSymbol { AssociatedSymbol: not null })
```

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:5193]: Consider asserting the expected symbol kinds here.

**Review by 333fred — APPROVED** [review]

**Comment by mavasani** [comment]
**Update:** After an offline discussion with @sharwell and @CyrusNajmabadi, we have concluded that changing the current local diagnostics semantics for accessor blocks will force us to re-analyze the entire property/indexer/event block for a single line edit within these symbols. This introduces an unnecessary performance overhead for lightbulb just to support a corner case scenario that was only hit by one analyzer. We also want to keep the functionality consistent between expression bodied and block bodied accessors, so just taking the fix for expression body case is also not preferable. So, we have decided to stick with the current driver semantics and not make any product changes here in the driver. Instead, we will update the analyzer to either report on `=>` token or wrap it's OperationBlock registration within a SymbolStart/End registration to explicitly specify broader analysis scope. I am going to revert all the product changes in this PR, and update the added tests in this PR to match the current semantics.

**Comment by mavasani** [comment]
FYI: I have reverted all the prior commits for this PR and pushed https://github.com/dotnet/roslyn/pull/67312/commits/1de40788b95f7b18b5d1b5a5e1faef7844d2be9a for the unit tests being added in the PR. I'll squash merge this after approval.

File: src/Compilers/CSharp/Test/Emit2/Diagnostics/GetDiagnosticsTests.cs
Comment (mavasani) [inline src/Compilers/CSharp/Test/Emit2/Diagnostics/GetDiagnosticsTests.cs:897]:
```suggestion
                    // Analyzers should not be allowed to report local diagnostics on the containing
```

**Review by 333fred — APPROVED** [review]

**Pull request merged**
